devices/src/legacy/i8042.rs

```diff
@@ -6,47 +6,256 @@
+const KEY_ALT: u16 = 0x0011;
+const KEY_DEL: u16 = 0xE071;
+
+/// Internal i8042 buffer size, in bytes
+const BUF_SIZE: usize = 16;
 
 /// A i8042 PS/2 controller that emulates just enough to shutdown the machine.
 pub struct I8042Device {
+    /// CPU reset eventfd. We will set this event when the guest issues CMD_RESET_CPU.
     reset_evt: EventFd,
+
+    /// Keyboard interrupt event (IRQ 1).
+    kbd_interrupt_evt: EventFd,
+
+    /// The i8042 status register.
+    status: u8,
+
+    /// The i8042 control register.
+    control: u8,
+
+    /// The i8042 output port.
+    outp: u8,
+
+    /// The last command sent to port 0x64.
+    cmd: u8,
+
+    /// The internal i8042 data buffer.
+    buf: [u8; BUF_SIZE],
+    buf_head: usize,
+    buf_tail: usize,
 }
 
 impl I8042Device {
-    /// Constructs a i8042 device that will signal the given event when the guest requests it.
-    pub fn new(reset_evt: EventFd) -> I8042Device {
-        I8042Device { reset_evt }
+    /// Constructs an i8042 device that will signal the given event when the guest requests it.
+    pub fn new(reset_evt: EventFd, kbd_interrupt_evt: EventFd) -> I8042Device {
+        I8042Device {
+            reset_evt,
+            kbd_interrupt_evt,
+            control: CB_POST_OK | CB_KBD_INT,
+            cmd: 0,
+            outp: 0,
+            status: SB_KBD_ENABLED,
+            buf: [0; BUF_SIZE],
+            buf_head: 0,
+            buf_tail: 0,
+        }
+    }
+
+    /// Returns a clone of the CPU reset event fd
+    pub fn get_reset_evt_clone(&self) -> Result<EventFd> {
+        self.reset_evt.try_clone().map_err(Error::CloneCpuResetEvt)
+    }
+
+    pub fn trigger_kbd_interrupt(&self) -> Result<()> {
+        if (self.control & CB_KBD_INT) == 0 {
+            return Err(Error::KbdInterruptDisabled);
+        }
+        self.kbd_interrupt_evt
+            .write(1)
+            .map_err(Error::KbdInterruptFailure)
+    }
+
+    pub fn trigger_key(&mut self, key: u16) -> Result<()> {
+        if key & 0xff00 != 0 {
+            self.push_byte((key >> 8) as u8);
+        }
+        self.push_byte((key & 0xff) as u8);
+
+        match self.trigger_kbd_interrupt() {
+            Ok(_) | Err(Error::KbdInterruptDisabled) => Ok(()),
+            Err(e) => Err(e),
+        }
+    }
+
+    pub fn trigger_ctrl_alt_del(&mut self) -> Result<()> {
+        self.trigger_key(KEY_CTRL)?;
+        self.trigger_key(KEY_ALT)?;
+        self.trigger_key(KEY_DEL)?;
+        Ok(())
+    }
+
+    fn push_byte(&mut self, byte: u8) {
+        self.status |= SB_OUT_DATA_AVAIL;
+        if (self.buf_head + 1) % BUF_SIZE == self.buf_tail {
+            self.buf_tail = (self.buf_tail + 1) % BUF_SIZE;
+        }
+        self.buf[self.buf_head] = byte;
+        self.buf_head = (self.buf_head + 1) % BUF_SIZE;
+    }
+
+    fn pop_byte(&mut self) -> Option<u8> {
+        if self.buf_head == self.buf_tail {
+            return None;
+        }
+
+        let res = self.buf[self.buf_tail];
```

Comment:
See the comment above. The actual capacity is 15 bytes, because I couldn't be bothered to correct my off-by-ones. In retrospect, this ring buffer was a bad idea, though. I'll update it with one that gets full and refuses pushes.